runtime/vm/stringhelpers.cpp

```diff
@@ -414,39 +414,40 @@ getStringUTF8Length(J9VMThread *vmThread, j9object_t string)
 UDATA 
 verifyQualifiedName(J9VMThread *vmThread, j9object_t string)
 {
+	J9JavaVM* vm = vmThread->javaVM;
 	UDATA unicodeLength = J9VMJAVALANGSTRING_LENGTH(vmThread, string);
-	j9object_t unicodeBytes = J9VMJAVALANGSTRING_VALUE(vmThread, string);
-	UDATA i = 0;
-	UDATA rc = CLASSNAME_VALID_NON_ARRARY;
+	char utf8NameStackBuffer[J9VM_PACKAGE_NAME_BUFFER_LENGTH];
+	J9CfrConstantPoolInfo classNameInfo = {0, 0, 0, 0, 0, NULL, 0};
+	IDATA arity = 0;
+
+	PORT_ACCESS_FROM_JAVAVM(vm);
 
-	if (unicodeLength == 0) {
+	if (0 == unicodeLength) {
 		return CLASSNAME_INVALID;
 	}
 
-	if (IS_STRING_COMPRESSED(vmThread, string)) {
-		if ('[' == J9JAVAARRAYOFBYTE_LOAD(vmThread, unicodeBytes, 0)) {
-			rc = CLASSNAME_VALID_ARRARY;
-		}
-		for (i = 0; i < unicodeLength; i++) {
-			U_8 c = J9JAVAARRAYOFBYTE_LOAD(vmThread, unicodeBytes, i);
-			if (c == '/') {
-				rc = CLASSNAME_INVALID;
-				break;
-			}
-		}
-	} else {
-		if ('[' == J9JAVAARRAYOFCHAR_LOAD(vmThread, unicodeBytes, 0)) {
-			rc = CLASSNAME_VALID_ARRARY;
-		}
-		for (i = 0; i < unicodeLength; i++) {
-			U_16 c = J9JAVAARRAYOFCHAR_LOAD(vmThread, unicodeBytes, i);
-			if (c == '/') {
-				rc = CLASSNAME_INVALID;
-				break;
-			}
-		}
+	classNameInfo.slot1 = (U_32)unicodeLength;
+	classNameInfo.bytes = (U_8*)copyStringToUTF8WithMemAlloc(vmThread, string,
+									J9_STR_NULL_TERMINATE_RESULT | J9_STR_XLAT, "", 0,
+									utf8NameStackBuffer, J9VM_PACKAGE_NAME_BUFFER_LENGTH, NULL);
+
+	/* keep checking other invalid characters in the class name */
+	arity = bcvCheckClassName(&classNameInfo);
+	if ((NULL != classNameInfo.bytes) && ((U_8*)utf8NameStackBuffer != classNameInfo.bytes)) {
+		j9mem_free_memory(classNameInfo.bytes);
+		classNameInfo.bytes = NULL;
 	}
-	return rc;
+
+	/* map the return code against the arity value from checkNameImpl (called by bcvCheckClassName):
+	 * 1) if arity = -1, then rc = CLASSNAME_INVALID(0) = arity + 1
+	 * 2) if arity = 0,  then rc = CLASSNAME_VALID_NON_ARRARY(1) = arity + 1
+	 * 3) if arity > 0,  then rc = CLASSNAME_VALID_ARRARY
+	 */
```

Comment:
I would rather we return the constant than +1 the arity.  This could as written will break if the constants are ever modified.

Reply:
Agreed and already replaced with these constants.